altair/datasets/_readers.py

```diff
@@ -0,0 +1,569 @@
+                depends = ", ".join(f"{req!r}" for req in reqs) + " packages"
+            else:
+                depends = f"{reqs!r} package"
+
+            msg = (
+                f"Backend {self._name!r} requires the {depends}, but {name!r} could not be found.\n"
+                f"This can be installed with pip using:\n"
+                f"    pip install {name}\n"
+                f"Or with conda using:\n"
+                f"    conda install -c conda-forge {name}"
+            )
+            raise ModuleNotFoundError(msg, name=name)
+
+    def __repr__(self) -> str:
+        return f"Reader[{self._name}]"
+
+    def __init__(self, name: LiteralString, /) -> None: ...
+
+
+class _PandasReader(_Reader["pd.DataFrame", "pd.DataFrame"]):
+    def __init__(self, name: _Pandas, /) -> None:
+        self._name = _requirements(name)
+        if not TYPE_CHECKING:
+            pd = self._import(self._name)
+        self._read_fn = {
+            ".csv": pd.read_csv,
+            ".json": pd.read_json,
+            ".tsv": partial["pd.DataFrame"](pd.read_csv, sep="\t"),
+            ".arrow": pd.read_feather,
+            ".parquet": pd.read_parquet,
+        }
+        self._scan_fn = {".parquet": pd.read_parquet}
+
+
+class _PandasPyArrowReader(_Reader["pd.DataFrame", "pd.DataFrame"]):
+    def __init__(self, name: Literal["pandas[pyarrow]"], /) -> None:
+        _pd, _pa = _requirements(name)
+        self._name = name
+        if not TYPE_CHECKING:
+            pd = self._import(_pd)
+            pa = self._import(_pa)  # noqa: F841
+
+        self._read_fn = {
+            ".csv": partial["pd.DataFrame"](pd.read_csv, dtype_backend=_pa),
+            ".json": partial["pd.DataFrame"](pd.read_json, dtype_backend=_pa),
+            ".tsv": partial["pd.DataFrame"](pd.read_csv, sep="\t", dtype_backend=_pa),
+            ".arrow": partial(pd.read_feather, dtype_backend=_pa),
+            ".parquet": partial(pd.read_parquet, dtype_backend=_pa),
+        }
+        self._scan_fn = {".parquet": partial(pd.read_parquet, dtype_backend=_pa)}
+
+
+def _pl_read_json_roundtrip(source: Path | IOBase, /, **kwds: Any) -> pl.DataFrame:
+    """
+    Try to utilize better date parsing available in `pl.read_csv`_.
+
+    `pl.read_json`_ has few options when compared to `pl.read_csv`_.
+
+    Chaining the two together - *where possible* - is still usually faster than `pandas.read_json`_.
+
+    .. _pl.read_json:
+        https://docs.pola.rs/api/python/stable/reference/api/polars.read_json.html
+    .. _pl.read_csv:
+        https://docs.pola.rs/api/python/stable/reference/api/polars.read_csv.html
+    .. _pandas.read_json:
+        https://pandas.pydata.org/docs/reference/api/pandas.read_json.html
+    """
+    from io import BytesIO
+
+    import polars as pl
+
+    df = pl.read_json(source, **kwds)
+    if any(tp.is_nested() for tp in df.schema.dtypes()):
+        # NOTE: Inferred as `(Geo|Topo)JSON`, which wouldn't be supported by `read_csv`
+        return df
```

Comment:
@joelostblom (https://github.com/vega/altair/pull/3631#issuecomment-2480816377)

> ### Non-dataframe data
> 
> What about the US10M and Worl_110M datasets? They are [both loaded as dictionaries rather than dataframes](https://github.com/altair-viz/vega_datasets/blob/136e850447b49031f04baa137ce5c37a6678bbb1/vega_datasets/core.py#L380-L405) in vega_datasets, so we would probably need to use special logic for those regardless I suppose.

`polars` can read all the datasets, including

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/tests/test_datasets.py#L404-L409


However, I'm pretty sure they won't work in a `alt.Chart` since they use nested data types

```py
from altair.datasets import Loader

data = Loader.with_backend("polars")
data("world-110m")

```

```
shape: (1, 4)
┌──────────┬───────────────────────┬──────────────────────────┬─────────────────────────────────┐
│ type     ┆ transform             ┆ objects                  ┆ arcs                            │
│ ---      ┆ ---                   ┆ ---                      ┆ ---                             │
│ str      ┆ struct[2]             ┆ struct[2]                ┆ list[list[list[i64]]]           │
╞══════════╪═══════════════════════╪══════════════════════════╪═════════════════════════════════╡
│ Topology ┆ {[0.0036,             ┆ {{"MultiPolygon",[[[0]], ┆ [[[33289, 2723], [-582, 81], …… │
│          ┆ 0.001693],[-180.0, -… ┆ [[1]]…                   ┆                                 │
└──────────┴───────────────────────┴──────────────────────────┴─────────────────────────────────┘

Reply:
I can assist with spatial data once needed. Use geopandas for pandas backend and geopolars/[polars-st](https://github.com/Oreilles/polars-st) for polars backend.
There are GEOS drivers available that can be used to read topojson and geojson structured data sources.

Reply:
> I can assist with spatial data once needed. Use geopandas for pandas backend and geopolars/[polars-st](https://github.com/Oreilles/polars-st) for polars backend. There are GEOS drivers available that can be used to read topojson and geojson structured data sources.

Perfect, I was hoping this part would interest you @mattijn 🎉.
I felt a little in-over-my-head on this.
There also appears to be https://github.com/geoarrow/geoarrow-rs, with ([python bindings](https://github.com/geoarrow/geoarrow-rs/blob/main/python/README.md)).

I was taking a look at [`geopandas.io.file`](https://github.com/geopandas/geopandas/blob/9b7e69459046cf8088720cca8b99fca98046c792/geopandas/io/file.py) to see if there was an established way to detect spatial json from regular json.
My hope was for cases like `pandas`, when the user doesn't have `geopandas`. Could we do something less heavy than attempting to read the file and erroring out?
I **prefer** [*look before you leap*](https://realpython.com/python-lbyl-vs-eafp/#the-look-before-you-leap-lbyl-style) over [*easier to ask forgiveness than permission*](https://realpython.com/python-lbyl-vs-eafp/#the-easier-to-ask-forgiveness-than-permission-eafp-style)

## Ideas
### Manually maintaining a record of all spatial data files 
- E.g. `earthquakes`, `world-110m`, `...`
- **Ideally** this would live in https://github.com/vega/vega-datasets
	- So that we don't get out of sync
	- But we *could* maintain this in `altair`
- I can add this as a column in [metadata.parquet](https://github.com/vega/altair/pull/3631#discussion_r1846456757)
	- Then we can dispatch to a different [read function](https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/altair/datasets/_readers.py#L99-L104) that correctly handles the data

### Reading *some* portion of the `.json` to infer
With `polars`, it is cheap enough to just read the whole thing and check the types (https://github.com/vega/altair/pull/3631#discussion_r1845659819).

For `pyarrow`, I'm *very loosely* approximating here.
But need to read the file into memory anyway, due to it only supporting [line-delimited JSON](https://arrow.apache.org/docs/python/json.html#reading-json-files)

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/altair/datasets/_readers.py#L400-L411

With `pandas`, I'm unsure what is best since it can read *most* `.json` files, but not spatial.
So I can't use the `polars` method (no schema/dtypes to inspect).
But also the `pyarrow` method will make all **non-spatial** `.json` much slower to read.

Reply:
Treat spatial support as an extension of the backend. If the backend is pandas and the file is known as a geojson or topojson file format use geopandas to read the file, so there will come a GeoInterface available (`.__geo_interface__`), since that will be used for serialization upon creating the chart.

For the polars backend use polars-st for now. They already use altair for their `.plot()` method, as can be seen here: https://github.com/Oreilles/polars-st/blob/main/polars_st/geodataframe.py#L453-L467 and that is possible since they provide a GeoInterface.

If a user uses pandas or polars for spatial data and has not geopandas or polars-st installed, report back that they should install that package. For other backends report back that the specific backend doesn't support spatial data. 

The URL should be available regardless the packages installed.

Generally speaking.
- For geopandas: 
```python
import geopandas as gpd
gdf = gpd.read_file(url)
```
- For polars-st
```python
import polars as pl
import polars_st as st
gdf = polars_st.read_file(url)
```
polars-st is still very new, so you also might opt to only use geopandas for now. In OP you mention that 5 backends have been implemented, but I cannot see which ones. So for now I assume we can provide spatial support to the pandas backend and potentially polars backend only. Which is great.